File: Python/service/src/service_nodes/service_nodes/service_sv_node.py
import rclpy
from rclpy.node import Node
# Gọi service vừa tạo ra
from add_two_ints.srv import SumTwoNums

class SrvServerNode(Node):
    def __init__(self):
        # Khởi tạo tên node, ở đây là tên service
        super().__init__('srv_server_node')

        # Tạo ra một service mới
        self.srv = self.create_service(
            SumTwoNums, #Loại service
            'sum_to_nums', # Đặt tên cho nó
            self.sum_to_nums_callback, # Gọi hàm callback
        )
    # Tạo callback

    def sum_to_nums_callback(self,request,response):
        # Xử lý callback:
        # - Nhận dữ liệu đầu vào là :  request.a và request.b
        # - Xử lý dữ liệu và phản hồi : response.sum
        response.sum = request.a + request.b

        #Debug dữ liệu
        self.get_logger().info(f'Request: {request.a} || {request.b} => Response: {response.sum}')

        # Trả về đối tượng response
        return response

def main(args=None):
    rclpy.init(args=args)

    srv_node = SrvServerNode()
    try:
        rclpy.spin(srv_node)
    except KeyboardInterrupt:
        srv_node.get_logger().info("This node is stopped by User !")
    finally:
        srv_node.destroy_node()
        rclpy.shutdown()

if __name__ == "__main__":
    main()

File: Cpp/service/src/cpp_srvcli/src/service_sv_node.cpp
#include "rclcpp/rclcpp.hpp"
#include "add_two_ints/srv/sum_two_nums.hpp"

#include <memory>

class SrvServerNode : public rclcpp::Node {
public:

    SrvServerNode() : Node("srv_server_node")
    {
        // Tạo máy chủ dịch vụ
        server_ = this->create_service<add_two_ints::srv::SumTwoNums>(
            "sum_two_nums", //Tên máy chủ của dịch vụ
            std::bind(
            &SrvServerNode::callbackSum, // Gọi callback để lắng nghe client
            this, // Trỏ chính đối tượng này
            std::placeholders::_1, // Đại diện cho tham số 1 (là request)
            std::placeholders::_2 // Đại diện cho tham số 2 (là response)
            )
        );

        RCLCPP_INFO(this->get_logger(), "Server is ready.");
    }

private:

    // callback when service request received
    void callbackSum(
        const std::shared_ptr<add_two_ints::srv::SumTwoNums::Request> request, // Tham số của request
        std::shared_ptr<add_two_ints::srv::SumTwoNums::Response> response // Tham số của response
    ){
        response->sum = request->a + request->b; // Lấy sum = a + b

        RCLCPP_INFO(
            this->get_logger(),
            "Incoming request: %ld + %ld = %ld",
            request->a, request->b, response->sum
        );
    }

    rclcpp::Service<add_two_ints::srv::SumTwoNums>::SharedPtr server_;
};


int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);

    auto node = std::make_shared<SrvServerNode>();

    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
